File: FWSegmentedControl_OC/FWSegmentedControl/ViewController.m
//
//  ViewController.m
//  FWSegmentedControl
//
//  Created by xfg on 2018/3/16.
//  Copyright © 2018年 xfg. All rights reserved.
//

#import "ViewController.h"
#import "FWSegmentedControl-Swift.h"
#import <FWSegmentedControl/FWSegmentedControl-Swift.h>
#import "ViewController2.h"

@interface ViewController () <UIScrollViewDelegate>

@property (nonatomic, strong) UIScrollView          *scrollView;                // ScrollView
@property (nonatomic, strong) NSArray               *sectionTitles3;

@property (nonatomic, assign) NSInteger             startPage;                  // Segmented的滑块起始页
@property (nonatomic, assign) BOOL                  isClickedSegmented;         // 是否点击了Segmented的滑块

@property (nonatomic, strong) FWSegmentedControl    *segmentedControl7;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.edgesForExtendedLayout = UIRectEdgeNone;
    self.view.backgroundColor = UIColor.lightGrayColor;
    self.navigationItem.title = @"FWSegmentedControl";
    
    NSArray *images2 = @[[UIImage imageNamed:@"a"], [UIImage imageNamed:@"b"], [UIImage imageNamed:@"c"]];
    
    NSArray *selectedImages2 = @[[UIImage imageNamed:@"a-selected"], [UIImage imageNamed:@"b-selected"], [UIImage imageNamed:@"c-selected"]];
    
    self.sectionTitles3 = @[@"福建", @"直播", @"小视频"];
    
    
    // 例一
    NSArray *sectionTitles = @[@"关注", @"游戏", @"附近", @"体育", @"女神范", @"运动啦啦", @"歌舞", @"吃鸡", @"户外", @"脱口秀"];
    FWSegmentedControl *segmentedControl = [FWSegmentedControl segmentedWithScType:SCTypeText scWidthStyle:SCWidthStyleDynamicFixedSuper sectionTitleArray:nil sectionImageArray:nil sectionSelectedImageArray:nil frame:CGRectMake(0, 0, self.view.frame.size.width, 50)];

    segmentedControl.sectionTitleArray = sectionTitles;
    segmentedControl.scSelectionIndicatorStyle = SCSelectionIndicatorStyleFullWidthStripe;

    [self.view addSubview:segmentedControl];



    // 例二
    FWSegmentedControl *segmentedControl2 = [FWSegmentedControl segmentedWithScType:SCTypeText scWidthStyle:SCWidthStyleDynamicFixedSuper sectionTitleArray:sectionTitles sectionImageArray:nil sectionSelectedImageArray:nil frame:CGRectMake(0, CGRectGetMaxY(segmentedControl.frame) + 10, self.view.frame.size.width, 40)];

    segmentedControl2.selectedSegmentIndex = 1;

    segmentedControl2.scSelectionIndicatorStyle = SCSelectionIndicatorStyleContentWidthStripe;
    segmentedControl2.scSelectionIndicatorLocation = SCSelectionIndicatorLocationDown;
    segmentedControl2.segmentEdgeInset = UIEdgeInsetsMake(0, 10, 0, 10);

    segmentedControl2.selectionIndicatorColor = UIColor.redColor;
    segmentedControl2.selectionIndicatorHeight = 3;

    segmentedControl2.titleTextAttributes = @{NSForegroundColorAttributeName: UIColor.grayColor, NSFontAttributeName: [UIFont systemFontOfSize:13.0]};
    segmentedControl2.selectedTitleTextAttributes = @{NSForegroundColorAttributeName: UIColor.redColor, NSFontAttributeName: [UIFont systemFontOfSize:13.0]};

    [self.view addSubview:segmentedControl2];



    // 例三
    NSArray *sectionTitles2 = @[@"女神", @"运动啦啦", @"歌舞"];

    FWSegmentedControl *segmentedControl3 = [FWSegmentedControl segmentedWithScType:SCTypeText scWidthStyle:SCWidthStyleDynamicFixedSuper sectionTitleArray:sectionTitles2 sectionImageArray:nil sectionSelectedImageArray:nil frame:CGRectMake(0, CGRectGetMaxY(segmentedControl2.frame) + 10, self.view.frame.size.width, 40)];

    segmentedControl3.scSelectionIndicatorStyle = SCSelectionIndicatorStyleBox;
    segmentedControl3.scSelectionIndicatorLocation = SCSelectionIndicatorLocationDown;
    segmentedControl3.segmentEdgeInset = UIEdgeInsetsMake(0, 10, 0, 10);

    segmentedControl3.selectionIndicatorColor = UIColor.redColor;
    segmentedControl3.selectionIndicatorHeight = 3;

    segmentedControl3.selectionIndicatorBoxColor = UIColor.greenColor;
    segmentedControl3.selectionIndicatorBoxOpacity = 0.4;

    segmentedControl3.verticalDividerEnabled = YES;
    segmentedControl3.verticalDividerColor = UIColor.lightGrayColor;
    segmentedControl3.verticalDividerWidth = 1.0;

    segmentedControl3.titleTextAttributes = @{NSForegroundColorAttributeName: UIColor.grayColor, NSFontAttributeName: [UIFont systemFontOfSize:13.0]};
    segmentedControl3.selectedTitleTextAttributes = @{NSForegroundColorAttributeName: UIColor.redColor, NSFontAttributeName: [UIFont systemFontOfSize:13.0]};

    [self.view addSubview:segmentedControl3];



    // 例四
    NSArray *images = @[[UIImage imageNamed:@"1"], [UIImage imageNamed:@"2"], [UIImage imageNamed:@"3"], [UIImage imageNamed:@"4"], [UIImage imageNamed:@"5"], [UIImage imageNamed:@"6"], [UIImage imageNamed:@"7"]];

    NSArray *selectedImages = @[[UIImage imageNamed:@"1-selected"], [UIImage imageNamed:@"2-selected"], [UIImage imageNamed:@"3-selected"], [UIImage imageNamed:@"4-selected"], [UIImage imageNamed:@"5-selected"], [UIImage imageNamed:@"6-selected"], [UIImage imageNamed:@"7-selected"]];

    FWSegmentedControl *segmentedControl4 = [FWSegmentedControl segmentedWithScType:SCTypeImages scWidthStyle:SCWidthStyleDynamicFixedSuper sectionTitleArray:nil sectionImageArray:images sectionSelectedImageArray:selectedImages frame:CGRectMake(0, CGRectGetMaxY(segmentedControl3.frame) + 10, self.view.frame.size.width, 40)];

    segmentedControl4.scSelectionIndicatorStyle = SCSelectionIndicatorStyleContentWidthStripe;
    segmentedControl4.scSelectionIndicatorLocation = SCSelectionIndicatorLocationUp;
    segmentedControl4.segmentEdgeInset = UIEdgeInsetsMake(0, 10, 0, 10);

    segmentedControl4.selectionIndicatorColor = UIColor.redColor;
    segmentedControl4.selectionIndicatorHeight = 5;

    segmentedControl4.verticalDividerEnabled = YES;
    segmentedControl4.verticalDividerColor = UIColor.lightGrayColor;
    segmentedControl4.verticalDividerWidth = 1.0;

    [self.view addSubview:segmentedControl4];


    // 例五
    NSDictionary *sectionSelectedImageDict = @{@(0) : [[FWSectionImageItem alloc] initWithItemImage:[UIImage imageNamed:@"hm_hot_city"] itemSelectedImage:[UIImage imageNamed:@"hm_hot_city_selected"]]};

    FWSegmentedControl *segmentedControl5 = [FWSegmentedControl segmentedWithScType:SCTypeTextImages scWidthStyle:SCWidthStyleDynamic sectionTitleArray:self.sectionTitles3 sectionSelectedImageDict:sectionSelectedImageDict frame:CGRectMake(0, CGRectGetMaxY(segmentedControl4.frame) + 10, self.view.frame.size.width, 40)];

    segmentedControl5.scSelectionIndicatorStyle = SCSelectionIndicatorStyleContentWidthStripe;
    segmentedControl5.scSelectionIndicatorLocation = SCSelectionIndicatorLocationDown;
    segmentedControl5.scImagePosition = SCImagePositionLeftOfText;
    segmentedControl5.segmentEdgeInset = UIEdgeInsetsMake(0, 10, 0, 10);

    segmentedControl5.selectionIndicatorColor = UIColor.redColor;
    segmentedControl5.selectionIndicatorHeight = 3;

    segmentedControl5.textImageSpacing = 4;

    segmentedControl5.titleTextAttributes = @{NSForegroundColorAttributeName: UIColor.grayColor, NSFontAttributeName: [UIFont systemFontOfSize:13.0]};
    segmentedControl5.selectedTitleTextAttributes = @{NSForegroundColorAttributeName: UIColor.redColor, NSFontAttributeName: [UIFont systemFontOfSize:13.0]};

    [self.view addSubview:segmentedControl5];


    // 例六
    FWSegmentedControl *segmentedControl6 = [FWSegmentedControl segmentedWithScType:SCTypeTextImages scWidthStyle:SCWidthStyleFixed sectionTitleArray:self.sectionTitles3 sectionImageArray:images2 sectionSelectedImageArray:selectedImages2 frame:CGRectMake(0, CGRectGetMaxY(segmentedControl5.frame) + 10, self.view.frame.size.width, 50)];

    segmentedControl6.scSelectionIndicatorStyle = SCSelectionIndicatorStyleArrowDown;
    segmentedControl6.scSelectionIndicatorLocation = SCSelectionIndicatorLocationDown;
    segmentedControl6.scImagePosition = SCImagePositionAboveText;
    segmentedControl6.segmentEdgeInset = UIEdgeInsetsMake(0, 10, 0, 10);

    segmentedControl6.selectionIndicatorColor = UIColor.redColor;
    segmentedControl6.selectionIndicatorHeight = 5;
    segmentedControl6.arrowWidth = 15.0;

    segmentedControl6.verticalDividerEnabled = YES;
    segmentedControl6.verticalDividerColor = UIColor.lightGrayColor;
    segmentedControl6.verticalDividerWidth = 1.0;

    segmentedControl6.textImageSpacing = 4;

    segmentedControl6.titleTextAttributes = @{NSForegroundColorAttributeName: UIColor.grayColor, NSFontAttributeName: [UIFont systemFontOfSize:13.0]};
    segmentedControl6.selectedTitleTextAttributes = @{NSForegroundColorAttributeName: UIColor.redColor, NSFontAttributeName: [UIFont systemFontOfSize:13.0]};

    [self.view addSubview:segmentedControl6];
    
    
    // 例七
    self.segmentedControl7 = [FWSegmentedControl segmentedWithScType:SCTypeTextImages scWidthStyle:SCWidthStyleDynamicFixedSuper sectionTitleArray:self.sectionTitles3 sectionImageArray:images2 sectionSelectedImageArray:selectedImages2 frame:CGRectMake(0, CGRectGetMaxY(segmentedControl6.frame) + 10, self.view.frame.size.width, 40)];

    self.segmentedControl7.scSelectionIndicatorStyle = SCSelectionIndicatorStyleFullWidthStripe;
    self.segmentedControl7.scSelectionIndicatorLocation = SCSelectionIndicatorLocationDown;
    self.segmentedControl7.scImagePosition = SCImagePositionRightOfText;
    self.segmentedControl7.segmentEdgeInset = UIEdgeInsetsMake(0, 10, 0, 10);

    self.segmentedControl7.selectionIndicatorColor = UIColor.redColor;
    self.segmentedControl7.selectionIndicatorHeight = 3;

    self.segmentedControl7.verticalDividerEnabled = YES;
    self.segmentedControl7.verticalDividerColor = UIColor.lightGrayColor;
    self.segmentedControl7.verticalDividerWidth = 1.0;

    self.segmentedControl7.textImageSpacing = 4;

    self.segmentedControl7.titleTextAttributes = @{NSForegroundColorAttributeName: UIColor.grayColor, NSFontAttributeName: [UIFont systemFontOfSize:13.0]};
    self.segmentedControl7.selectedTitleTextAttributes = @{NSForegroundColorAttributeName: UIColor.redColor, NSFontAttributeName: [UIFont systemFontOfSize:13.0]};

    __weak typeof(self) ws = self;
    self.segmentedControl7.indexChangeBlock = ^(NSInteger index) {

        ws.isClickedSegmented = YES;
        [ws.scrollView scrollRectToVisible:CGRectMake(ws.view.frame.size.width * index, 0, ws.view.frame.size.width, CGRectGetHeight(ws.scrollView.frame)) animated:YES];
    };

    [self.view addSubview:self.segmentedControl7];
    
    
    self.scrollView = [[UIScrollView alloc] initWithFrame:CGRectMake(0, CGRectGetMaxY(self.segmentedControl7.frame), self.view.frame.size.width, self.view.frame.size.height - CGRectGetMaxY(self.segmentedControl7.frame))];
    self.scrollView.delegate = self;
    self.scrollView.backgroundColor = UIColor.clearColor;
    self.scrollView.pagingEnabled = YES;
    self.scrollView.showsHorizontalScrollIndicator = NO;
    self.scrollView.bounces = NO;
    self.scrollView.contentSize = CGSizeMake(self.view.frame.size.width * [self.sectionTitles3 count], CGRectGetHeight(self.scrollView.frame));
    [self.scrollView scrollRectToVisible:CGRectMake(0, 0, self.view.frame.size.width, CGRectGetHeight(self.scrollView.frame)) animated:NO];
    [self.view addSubview:self.scrollView];
    
    [self.scrollView addSubview:[self setupUIView:0]];
    [self.scrollView addSubview:[self setupUIView:1]];
    [self.scrollView addSubview:[self setupUIView:2]];
    
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"多级联动>" style:UIBarButtonItemStylePlain target:self action:@selector(goNextVC)];
}

- (UILabel *)setupUIView:(int)index
{
    UILabel *label = [[UILabel alloc] initWithFrame:CGRectMake(self.view.frame.size.width * index, 0, self.view.frame.size.width, CGRectGetHeight(self.scrollView.frame))];
    label.text = self.sectionTitles3[index];
    label.textColor = UIColor.whiteColor;
    label.textAlignment = NSTextAlignmentCenter;
    label.backgroundColor = [UIColor colorWithRed:(float)(1+arc4random()%99)/100 green:(float)(1+arc4random()%99)/100 blue:(float)(1+arc4random()%99)/100 alpha:1];
    return label;
}

- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    if (!self.isClickedSegmented)
    {
        CGFloat pageWidth = scrollView.frame.size.width;
        NSInteger tmpPage = scrollView.contentOffset.x / pageWidth;
        float tmpPage2 = scrollView.contentOffset.x / pageWidth;
        NSInteger page = tmpPage2-tmpPage>=0.5 ? tmpPage+1 : tmpPage;
        
        if (_startPage != page)
        {
            [self.segmentedControl7 setSelectedSegmentIndexWithIndex:page animated:YES];
            _startPage = page;
        }
    }
}

- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView
{
    CGFloat pageWidth = scrollView.frame.size.width;
    NSInteger page = scrollView.contentOffset.x / pageWidth;
    
    self.isClickedSegmented = NO;
    
    [self.segmentedControl7 setSelectedSegmentIndexWithIndex:page animated:YES];
}

- (void)goNextVC
{
    [self.navigationController pushViewController:[[ViewController2 alloc] init] animated:YES];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: FWSegmentedControl/FWSegmentedControl/ViewController.swift
//
//  ViewController.swift
//  FWSegmentedControl
//
//  Created by xfg on 2018/3/12.
//  Copyright © 2018年 xfg. All rights reserved.
//

import UIKit

/// segmentedControl字体大小
let kSegmentTitleFont: CGFloat = 13.0
/// segmentedControl左边边距
let kSegmentLeftEdge: CGFloat = 10

class ViewController: UIViewController, UIScrollViewDelegate {
    
    /// 是否点击了Segmented的滑块
    var isClickedSegmented = false
    /// Segmented的滑块起始页
    var startPage = 0
    
    let sectionTitles = ["关注", "游戏", "附近", "体育", "女神范", "运动啦啦", "歌舞", "吃鸡", "户外", "脱口秀"]
    
    let sectionTitles2 = ["女神", "运动啦啦", "歌舞"]
    
    let images = [UIImage(named: "1"),
                  UIImage(named: "2"),
                  UIImage(named: "3"),
                  UIImage(named: "4"),
                  UIImage(named: "5"),
                  UIImage(named: "6"),
                  UIImage(named: "7")]
    
    let selectedImages = [UIImage(named: "1-selected"),
                          UIImage(named: "2-selected"),
                          UIImage(named: "3-selected"),
                          UIImage(named: "4-selected"),
                          UIImage(named: "5-selected"),
                          UIImage(named: "6-selected"),
                          UIImage(named: "7-selected")]
    
    let imageTitles3 = ["叶子", "椰子啦", "叶紫"]
    
    let images2 = [UIImage(named: "a"),
                  UIImage(named: "b"),
                  UIImage(named: "c")]
    
    let selectedImages2 = [UIImage(named: "a-selected"),
                          UIImage(named: "b-selected"),
                          UIImage(named: "c-selected")]
    
    /// 例一
    private lazy var segmentedControl: FWSegmentedControl = {
        
        let segmentedControl = FWSegmentedControl.segmentedWith(scType: SCType.text, scWidthStyle: SCWidthStyle.fixed, sectionTitleArray: nil, sectionImageArray: nil, sectionSelectedImageArray: nil, frame: CGRect(x: 0, y: 0, width: Int(UIScreen.main.bounds.width), height: 50))
        
        segmentedControl.sectionTitleArray = sectionTitles
        segmentedControl.scSelectionIndicatorStyle = .fullWidthStripe
        
        return segmentedControl
    }()
    
    /// 例二
    private lazy var segmentedControl2: FWSegmentedControl = {
        
        let segmentedControl = FWSegmentedControl.segmentedWith(scType: SCType.text, scWidthStyle: SCWidthStyle.dynamicFixedSuper, sectionTitleArray: sectionTitles, sectionImageArray: nil, sectionSelectedImageArray: nil, frame: CGRect(x: 0, y: Int(self.segmentedControl.frame.maxY) + 10, width: Int(UIScreen.main.bounds.width), height: 40))
        
        segmentedControl.selectedSegmentIndex = 1
        
        segmentedControl.scSelectionIndicatorStyle = .contentWidthStripe
        segmentedControl.segmentEdgeInset = UIEdgeInsets(top: 0, left: kSegmentLeftEdge, bottom: 0, right: kSegmentLeftEdge)
        
        segmentedControl.selectionIndicatorColor = UIColor.red
        segmentedControl.selectionIndicatorHeight = 3
        
        segmentedControl.titleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.gray, NSAttributedStringKey.font : UIFont.systemFont(ofSize: kSegmentTitleFont)]
        segmentedControl.selectedTitleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.red, NSAttributedStringKey.backgroundColor: UIColor.clear, NSAttributedStringKey.font: UIFont.systemFont(ofSize: kSegmentTitleFont)]
        
        segmentedControl.autoresizingMask = [.flexibleRightMargin, .flexibleLeftMargin]
        
        return segmentedControl
    }()
    
    /// 例三
    private lazy var segmentedControl3: FWSegmentedControl = {
        
        let segmentedControl = FWSegmentedControl.segmentedWith(scType: SCType.text, scWidthStyle: SCWidthStyle.fixed, sectionTitleArray: sectionTitles2, sectionImageArray: nil, sectionSelectedImageArray: nil, frame: CGRect(x: 0, y: Int(self.segmentedControl2.frame.maxY) + 10, width: Int(UIScreen.main.bounds.width), height: 40))
        
        segmentedControl.backgroundColor = UIColor.clear
        segmentedControl.scSelectionIndicatorStyle = .box
        segmentedControl.scSelectionIndicatorLocation = .down
        segmentedControl.segmentEdgeInset = UIEdgeInsets(top: 0, left: kSegmentLeftEdge, bottom: 0, right: kSegmentLeftEdge)
        
        segmentedControl.selectionIndicatorBoxColor = UIColor.green
        segmentedControl.selectionIndicatorBoxOpacity = 0.4
        segmentedControl.selectionIndicatorBoxFollowEdgeInset = true
        segmentedControl.selectionIndicatorHeight = 0
        
        segmentedControl.scBorderType = [.left, .bottom, .top, .right]
        segmentedControl.segmentBorderColor = UIColor.red
        segmentedControl.segmentBorderWidth = 2.0
        segmentedControl.segmentBackgroundColor = UIColor.white
        
        segmentedControl.titleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.gray, NSAttributedStringKey.font : UIFont.systemFont(ofSize: kSegmentTitleFont)]
        segmentedControl.selectedTitleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.red, NSAttributedStringKey.backgroundColor: UIColor.clear, NSAttributedStringKey.font: UIFont.systemFont(ofSize: kSegmentTitleFont)]
        
        segmentedControl.autoresizingMask = .flexibleRightMargin
        
        return segmentedControl
    }()
    
    /// 例四
    private lazy var segmentedControl4: FWSegmentedControl = {
        
        let segmentedControl = FWSegmentedControl.segmentedWith(scType: SCType.text, scWidthStyle: SCWidthStyle.dynamic, sectionTitleArray: sectionTitles, sectionImageArray: nil, sectionSelectedImageArray: nil, frame: CGRect(x: 0, y: Int(self.segmentedControl3.frame.maxY) + 10, width: Int(UIScreen.main.bounds.width), height: 30))
        
        segmentedControl.backgroundColor = UIColor.clear
        segmentedControl.scSelectionIndicatorStyle = .box
        segmentedControl.scSelectionIndicatorLocation = .down
        segmentedControl.segmentEdgeInset = UIEdgeInsets(top: 0, left: 6.0, bottom: 0, right: 6.0)
        segmentedControl.segmentBackgroundEdgeInset = UIEdgeInsets(top: 0, left: 15, bottom: 0, right: 15)
        
        segmentedControl.selectionIndicatorBoxColor = UIColor.red
        segmentedControl.selectionIndicatorBoxEdgeInset = UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10)
        segmentedControl.selectionIndicatorBoxFollowEdgeInset = true
        segmentedControl.selectionIndicatorHeight = 0
        
        segmentedControl.enlargetouchesEdgeInset = UIEdgeInsets(top: 5, left: 5, bottom: 5, right: 5)
        
        segmentedControl.segmentBackgroundCornerRadius = 15
        segmentedControl.segmentBackgroundColor = UIColor.white
        
        segmentedControl.titleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.black, NSAttributedStringKey.font : UIFont.systemFont(ofSize: kSegmentTitleFont)]
        segmentedControl.selectedTitleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.white, NSAttributedStringKey.backgroundColor: UIColor.clear, NSAttributedStringKey.font: UIFont.systemFont(ofSize: kSegmentTitleFont)]
        
        segmentedControl.autoresizingMask = .flexibleRightMargin
        
        return segmentedControl
    }()
    
    /// 例五
    private lazy var segmentedControl5: FWSegmentedControl = {
        
        let segmentedControl = FWSegmentedControl.segmentedWith(scType: SCType.images, scWidthStyle: SCWidthStyle.dynamicFixedSuper, sectionTitleArray: nil, sectionImageArray: images as? [UIImage], sectionSelectedImageArray: selectedImages as? [UIImage], frame: CGRect(x: 0, y: Int(self.segmentedControl4.frame.maxY) + 10, width: Int(UIScreen.main.bounds.width), height: 40))
        
        segmentedControl.scSelectionIndicatorStyle = .contentWidthStripe
        segmentedControl.scSelectionIndicatorLocation = .up
        segmentedControl.segmentEdgeInset = UIEdgeInsets(top: 0, left: kSegmentLeftEdge, bottom: 0, right: kSegmentLeftEdge)
        
        segmentedControl.selectionIndicatorColor = UIColor.red
        segmentedControl.selectionIndicatorHeight = 5
        
        segmentedControl.verticalDividerEnabled = true
        segmentedControl.verticalDividerColor = UIColor.lightGray
        segmentedControl.verticalDividerWidth = 1.0
        
        segmentedControl.autoresizingMask = .flexibleRightMargin
        
        return segmentedControl
    }()
    
    /// 例六
    private lazy var segmentedControl6: FWSegmentedControl = {
        
        let imageTitles5 = ["福建", "直播", "小视频", "", "附近"]
        
        let sectionSelectedImageDict = [
            0 : FWSectionImageItem(itemImage: UIImage(named: "hm_hot_city")!, itemSelectedImage: UIImage(named: "hm_hot_city_selected")),
            3 : FWSectionImageItem(itemImage: UIImage(named: "a")!, itemSelectedImage: UIImage(named: "a-selected")),
        ]
        
        let segmentedControl = FWSegmentedControl.segmentedWith(scType: SCType.textImages, scWidthStyle: SCWidthStyle.dynamic, sectionTitleArray: imageTitles5, sectionSelectedImageDict: sectionSelectedImageDict, frame: CGRect(x: 0, y: Int(self.segmentedControl5.frame.maxY) + 10, width: Int(UIScreen.main.bounds.width), height: 40))
        
        segmentedControl.scSelectionIndicatorStyle = .contentWidthStripe
        segmentedControl.scSelectionIndicatorLocation = .down
        segmentedControl.scImagePosition = .rightOfText
        segmentedControl.segmentEdgeInset = UIEdgeInsets(top: 0, left: kSegmentLeftEdge, bottom: 0, right: kSegmentLeftEdge)
        
        segmentedControl.selectionIndicatorColor = UIColor.red
        segmentedControl.selectionIndicatorHeight = 3
        segmentedControl.selectionIndicatorEdgeInsets = UIEdgeInsets(top: 0, left: 5, bottom: 0, right: 5)
        segmentedControl.selectionIndicatorFollowText = true
        
        segmentedControl.textImageSpacing = 4
        
        segmentedControl.titleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.gray, NSAttributedStringKey.font : UIFont.systemFont(ofSize: kSegmentTitleFont)]
        segmentedControl.selectedTitleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.red, NSAttributedStringKey.backgroundColor: UIColor.clear, NSAttributedStringKey.font: UIFont.systemFont(ofSize: kSegmentTitleFont)]
        
        segmentedControl.autoresizingMask = .flexibleRightMargin
        
        return segmentedControl
    }()
    
    /// 例七
    private lazy var segmentedControl7: FWSegmentedControl = {
        
        let segmentedControl = FWSegmentedControl.segmentedWith(scType: SCType.textImages, scWidthStyle: SCWidthStyle.fixed, sectionTitleArray: imageTitles3, sectionImageArray: images2 as? [UIImage], sectionSelectedImageArray: selectedImages2 as? [UIImage], frame: CGRect(x: 0, y: Int(self.segmentedControl6.frame.maxY) + 10, width: Int(UIScreen.main.bounds.width), height: 50))
        
        segmentedControl.scSelectionIndicatorStyle = .arrowDown
        segmentedControl.scSelectionIndicatorLocation = .down
        segmentedControl.scImagePosition = .aboveText
        segmentedControl.segmentEdgeInset = UIEdgeInsets(top: 0, left: kSegmentLeftEdge, bottom: 0, right: kSegmentLeftEdge)
        
        segmentedControl.selectionIndicatorColor = UIColor.red
        segmentedControl.selectionIndicatorHeight = 5
        
        segmentedControl.verticalDividerEnabled = true
        segmentedControl.verticalDividerColor = UIColor.lightGray
        segmentedControl.verticalDividerWidth = 2.0
        
        segmentedControl.textImageSpacing = 2.0
        
        segmentedControl.arrowWidth = 15
        
        segmentedControl.titleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.gray, NSAttributedStringKey.font : UIFont.systemFont(ofSize: kSegmentTitleFont)]
        segmentedControl.selectedTitleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.red, NSAttributedStringKey.backgroundColor: UIColor.clear, NSAttributedStringKey.font: UIFont.systemFont(ofSize: kSegmentTitleFont)]
        
        segmentedControl.autoresizingMask = .flexibleRightMargin
        
        return segmentedControl
    }()
    
    /// 例八
    private lazy var segmentedControl8: FWSegmentedControl = {
        
        let segmentedControl = FWSegmentedControl.segmentedWith(scType: SCType.textImages, scWidthStyle: SCWidthStyle.dynamicFixedSuper, sectionTitleArray: imageTitles3, sectionImageArray: images2 as? [UIImage], sectionSelectedImageArray: selectedImages2 as? [UIImage], frame: CGRect(x: 0, y: Int(self.segmentedControl7.frame.maxY) + 10, width: Int(UIScreen.main.bounds.width), height: 40))
        
        segmentedControl.scSelectionIndicatorStyle = .fullWidthStripe
        segmentedControl.scSelectionIndicatorLocation = .down
        segmentedControl.scImagePosition = .rightOfText
        segmentedControl.segmentEdgeInset = UIEdgeInsets(top: 0, left: kSegmentLeftEdge, bottom: 0, right: kSegmentLeftEdge)
        
        segmentedControl.selectionIndicatorColor = UIColor.red
        segmentedControl.selectionIndicatorHeight = 3
        
        segmentedControl.verticalDividerEnabled = true
        segmentedControl.verticalDividerColor = UIColor.lightGray
        segmentedControl.verticalDividerWidth = 1.0
        
        segmentedControl.textImageSpacing = 4
        
        segmentedControl.titleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.gray, NSAttributedStringKey.font : UIFont.systemFont(ofSize: kSegmentTitleFont)]
        segmentedControl.selectedTitleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.red, NSAttributedStringKey.backgroundColor: UIColor.clear, NSAttributedStringKey.font: UIFont.systemFont(ofSize: kSegmentTitleFont)]
        
        segmentedControl.autoresizingMask = .flexibleRightMargin
        
        return segmentedControl
    }()
    
    lazy var scrollView: UIScrollView = {
        
        let scrollView = UIScrollView(frame: CGRect(x: 0, y: self.segmentedControl8.frame.maxY, width: self.view.frame.width, height: self.view.frame.height-self.segmentedControl8.frame.maxY))
        scrollView.backgroundColor = UIColor.clear
        scrollView.delegate = self
        scrollView.isPagingEnabled = true
        scrollView.showsHorizontalScrollIndicator = false
        scrollView.bounces = false
        scrollView.contentSize = CGSize(width: self.view.bounds.width * CGFloat(imageTitles3.count), height: self.view.bounds.height-self.segmentedControl8.frame.maxY)
        scrollView.scrollRectToVisible(CGRect(x: 0, y: 0, width: self.view.bounds.width, height: scrollView.frame.height), animated: false)
        return scrollView
    }()

    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.backgroundColor = UIColor.lightGray
        self.navigationItem.title = "FWSegmentedControl"
        self.edgesForExtendedLayout = []
        
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "多级联动>", style: .plain, target: self, action: #selector(goNextVC))
        
        self.view.addSubview(self.segmentedControl)
        self.view.addSubview(self.segmentedControl2)
        self.view.addSubview(self.segmentedControl3)
        self.view.addSubview(self.segmentedControl4)
        self.view.addSubview(self.segmentedControl5)
        self.view.addSubview(self.segmentedControl6)
        self.view.addSubview(self.segmentedControl7)
        self.view.addSubview(self.segmentedControl8)
        self.view.addSubview(self.scrollView)

        self.segmentedControl8.indexChangeBlock = { [weak self] index in
            self?.isClickedSegmented = true
            self?.scrollView.scrollRectToVisible(CGRect(x: (self?.view.bounds.width)! * CGFloat(index), y: 0, width: (self?.view.bounds.width)!, height: (self?.scrollView.frame.height)!), animated: true)
        }

        self.scrollView.addSubview(self.setupUIView(index: 0))
        self.scrollView.addSubview(self.setupUIView(index: 1))
        self.scrollView.addSubview(self.setupUIView(index: 2))
    }
}

extension ViewController {
    
    func setupUIView(index: Int) -> UILabel {
        
        let label = UILabel(frame: CGRect(x: self.view.frame.width * CGFloat(index), y: 0, width: self.view.frame.width, height: self.scrollView.frame.height))
        label.backgroundColor = UIColor(red: CGFloat(arc4random()%256)/255.0, green: CGFloat(arc4random()%256)/255.0, blue: CGFloat(arc4random()%256)/255.0, alpha: 1.0)
        label.textColor = UIColor.white
        label.text = self.imageTitles3[index]
        label.textAlignment = .center
        return label
    }
    
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        
        if isClickedSegmented == false {
            let pageWidth = scrollView.frame.width
            let tmpPage = scrollView.contentOffset.x / pageWidth
            let tmpPage2 = scrollView.contentOffset.x / pageWidth
            let page = tmpPage2-tmpPage>=0.5 ? tmpPage+1 : tmpPage
            if startPage != Int(page) {
                self.segmentedControl8.setSelectedSegmentIndex(index: Int(page), animated: true)
                startPage = Int(page)
            }
        }
    }
    
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        let pageWidth = scrollView.frame.width
        let page = scrollView.contentOffset.x / pageWidth
        self.isClickedSegmented = false
        self.segmentedControl8.setSelectedSegmentIndex(index: Int(page), animated: true)
    }
    
    @objc func goNextVC() {
        
        self.navigationController?.pushViewController(ViewController2(), animated: true)
    }
}

